### PR TITLE
Fix ReduceLROnPlateau's patience mechanic.

### PR DESCRIPTION
**Pull request by zou3519**
Let's say patience = 1 and the losses are:
[1.0, 1.1, 1.2, ..., ] (we've hit a plateau). Let's also say the losses
each come at one unit per time (at the `t = 0`, we receive losses[0], etc).

`self.patience` is 1.

at `t=0`:
- `self.num_bad_epochs` is 0.
- The epoch is better than the last one (which has a "loss value" of
  `self.mode_worst`. `self.num_bad_epochs` gets set to 0.

at `t=1`:
- `self.num_bad_epochs` is 0.
- The epoch is not better than the last one (which had a "loss value" of
  `1.0`). `self.num_bad_epochs` gets set to 1.
- The LR isn't reduced, because the check
  (`self.num_bad_epochs > self.patience`) is false

at `t=2`
- `self.num_bad_epochs` is 0.
- The epoch is not better than the last one (which had a "loss value" of
  `1.1`). `self.num_bad_epochs` gets set to 2.
- The LR is reduced, because the check (`self.num_bad_epochs > self.patience`)
  is true (2 > 1)

However, I would expect that at `t=1` the LR is reduced because, in the
words of the documentation, "Number of epochs with no improvement after
which learning rate will be reduced" implies that with `patience = 1`,
after one epoch with no improvement, the LR should be reduced.

Bug was reported by @W4ngatang.

If we thing this fix is good (this might break user code, so it might be better to just change our documentation?) I'll add a test. I'm not sure what the best way to test this is; I could try mocking the optimizer.

### CODE REVIEW

**Comment by apaszke** [comment]
I think all boils down to how "patience" is understood by different people. Consider those two definitions:
- How many bad epochs can you see before you start considering changing the learning rate.
- How many bad epochs can you see before you change the learning rate.

The first point gives you the current code, the second one gives the one after your patch. I have no strong preference towards any of them, although I agree that the second one might seem simpler.

Regarding the tests, we do have some for this feature, so you'll need to patch those up.

**Comment by zou3519** [comment]
thank you for the explanation, @apaszke!

I did some snooping around and our `ReduceLROnPlateau` implementation is ported from Keras (https://github.com/pytorch/pytorch/pull/1370). The current implementation (without this patch) behaves the same way as Keras' `ReduceLROnPlateau` so I think I won't change its behavior for consistency and instead update the docs

**Comment by apaszke** [comment]
Right. In that situation, we should really keep it as it is right now.
